src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -194,6 +196,45 @@ public static bool IsReferenceOrContainsReferences<T>()
             // See getILIntrinsicImplementation for how this happens.
             throw new InvalidOperationException();
         }
+
+        // Returns true iff the object has a component size;
+        // i.e., is variable length like string, array, Utf8String.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe bool ObjectHasComponentSize(object obj)
+        {
+            // CLR objects are laid out in memory as follows.
+            // [ pMethodTable || .. object data .. ]
+            //   ^-- the object reference points here
+            //
+            // The first DWORD of the method table class will have its high bit set if the
+            // method table has component size info stored somewhere. See member
+            // MethodTable:IsStringOrArray in src\vm\methodtable.h for full details.
+            //
+            // So in effect this method is the equivalent of
+            // return ((MethodTable*)(*obj))->IsStringOrArray();
+
+            Debug.Assert(obj != null);
+            return *(int*)GetObjectMethodTablePointer(obj) < 0;
+        }
+
+        // Given an object reference, returns its MethodTable* as an IntPtr.
+        //[Intrinsic]
+        //[NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static IntPtr GetObjectMethodTablePointer(object obj)
+        {
+            Debug.Assert(obj != null);
+
+            // We know that the first data field in any managed object is immediately after the
+            // method table pointer, so just back up one pointer and immediately deref.
+            // This is not ideal in terms of minimizing instruction count but is the best we can do at the moment.
+
+            return Unsafe.Add(ref Unsafe.As<byte, IntPtr>(ref JitHelpers.GetPinningHelper(obj).m_data), -1);
+
+            // Ideally this method would be replaced by the VM with:
```

Comment:
I do not think we would ever want this. This would mean hacking the JIT to allow derefencing objectrefs as pointers. It is invalid IL and it should stay as invalid IL.

I think what we would want here is to make methodtable pointer to be a field and fetch the field here, similar to how it is done in CoreRT.

Reply:
As an experiment, I tried writing an `[Intrinsic]` method that was VM-replaced by the mentioned IL sequence. The release VM and JIT handled it just fine (and produced the desired codegen), but the checked JIT hit an assert and failed. I'm sure other parts of the JIT would get angry if I tried to force this through. :)

Really all this gets us in the end is that we eliminate a single instruction. It allows us to turn this (the codegen from this PR):

```asm
lea rax, [rbx + 8] ; rbx = obj ref, rax = ref obj._firstField
mov rax, qword ptr [rax - 8] ; rax = pMethodTable
; .. remainder of logic ..
```

Into this (note no `lea` instruction or SIB syntax):

```asm
mov rax, qword ptr [rbx] ; rbx = obj ref, rax = pMethodTable
; .. remainder of logic ..
```

Reply:
> the checked JIT hit an assert and failed. I'm sure other parts of the JIT would get angry if I tried to force this through. :)

Right, it is invalid IL and it should stay invalid IL. The comment that says we would want to ideally replace it by `ldarg.0` + `ldind.i` should be deleted.

Reply:
The right fix for this should be on the VM side, not in the JIT. To make this work, VM should feed the following instruction to the JIT: `ldarg.0` + `ldfld m_pMethodTable`.

Reply:
@jkotas - But would it be incorrect for the JIT to fold this? I know that we can get into trouble if we incorrectly optimize address expressions to produce interim byref results that don't point into an object, but it would seem that folding in this case could never be invalid - perhaps I'm missing something?

Reply:
> But would it be incorrect for the JIT to fold this?

Yes, folding the two instructions in the JIT should be fine too.